dbusmock/templates/bluez5.py

```diff
@@ -393,3 +498,625 @@ def DisconnectDevice(_self, adapter_device_name, device_address):
+                              191])
+
+
+@dbus.service.method(BLUEZ_MOCK_IFACE,
+                     in_signature='ss', out_signature='')
+def ConnectMicroBit(self, adapter_name, device_address):
+    print('In connect microbit')
+    upper_address = device_address.upper().replace(":", "_")
+    dev_path = f'/org/bluez/{adapter_name}/dev_{upper_address}'
+    device = mockobject.objects[dev_path]
+
+    self.ConnectDevice(adapter_name, upper_address)
+
+    for path in microbit_data:
+        srvc_props = microbit_data[path].get(GATT_SRVC_IFACE)
+        if srvc_props:
+            del srvc_props['Includes']
+            print('adding service', srvc_props)
+            self.AddGattService(dbus.ObjectPath(path),
+                                dbus.Dictionary(srvc_props,
+                                                signature='sv'))
+        chrc_props = microbit_data[path].get(GATT_CHRC_IFACE)
+        if chrc_props:
+            print('add characteristic')
+            chrc_props = dbus.Dictionary(chrc_props, signature='sv')
+            chrc_props['Value'] = dbus.Array(chrc_props['Value'], signature='y')
+            self.AddGattCharacteristic(dbus.ObjectPath(path), chrc_props)
+
+    device.props[DEVICE_IFACE]['ServicesResolved'] = dbus.Boolean(True, variant_level=1)
+
+    device.EmitSignal(dbus.PROPERTIES_IFACE, 'PropertiesChanged', 'sa{sv}as', [
+        DEVICE_IFACE,
+        {
+            'ServicesResolved': dbus.Boolean(True, variant_level=1),
+        },
+        [],
+    ])
+
+
+@dbus.service.method(BLUEZ_MOCK_IFACE,
+                     in_signature='oa{sv}', out_signature='ay')
+def GattReadValue(self, path, options):
+    gatt_chrc = mockobject.objects[path]
+    gatt_chrc.call_log.append((int(time.time()), 'ReadValue', [options]))
+    # return microbit_data[path].get(GATT_CHRC_IFACE, {}).get('Value')
+    return gatt_chrc.Get(GATT_CHRC_IFACE, 'Value')
+
+
+@dbus.service.method(BLUEZ_MOCK_IFACE,
+                     in_signature='oaya{sv}', out_signature='')
+def GattWriteValue(self, path, value, options):
+    # microbit_data[path][GATT_CHRC_IFACE]['Value'] = dbus.Array(value)
+    gatt_chrc = mockobject.objects[path]
+    gatt_chrc.Set(GATT_CHRC_IFACE, 'Value', value)
+    gatt_chrc.call_log.append((int(time.time()), 'WriteValue', [value, options]))
+    if path == '/org/bluez/hci0/dev_DD_02_02_02_02_02/service0028/char0029':
+        tx_path = '/org/bluez/hci0/dev_DD_02_02_02_02_02/service0028/char002b'
+        tx_obj = mockobject.objects[tx_path]
+        tx_obj.EmitSignal(dbus.PROPERTIES_IFACE,
+                          'PropertiesChanged',
+                          'sa{sv}as', [
+                              GATT_CHRC_IFACE,
+                              {
+                                  'Value': dbus.Array(value, variant_level=1),
+                              },
+                              [],
+                          ])
+
+
+microbit_data = {
```

Comment:
This really needs to be done some other way...

Reply:
Seconded, please keep out this giant blurb of magic numbers. This really belongs into the test suite that *uses* the dbusmock. Note that these are *templates*, and test suites are expect to configure/modify it as required for a particular test. This should provide a skeleton, not a set of highly specific behaviour.